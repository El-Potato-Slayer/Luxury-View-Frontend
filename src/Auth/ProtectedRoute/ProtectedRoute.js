import PropTypes from 'prop-types';
import { useSelector } from 'react-redux';
import { Redirect, Route } from 'react-router-dom';

function ProtectedRoute({
  component: Component, exact, path,
}) {
  const { isLoggedIn } = useSelector((state) => state.userReducer);
  return (
    <Route
      exact={exact}
      path={path}
      render={(props) => {
        if (isLoggedIn) {
          return <Component />;
        }
        return <Redirect to={{ pathname: '/login', state: { from: props.location } }} />;
      }}
    />
  );
}

ProtectedRoute.propTypes = {
  component: PropTypes.instanceOf(Object).isRequired,
  exact: PropTypes.bool,
  location: PropTypes.instanceOf(Object),
  path: PropTypes.string.isRequired,
};

ProtectedRoute.defaultProps = {
  exact: false,
  location: {},
};

export default ProtectedRoute;
