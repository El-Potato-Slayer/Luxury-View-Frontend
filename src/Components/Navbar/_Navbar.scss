.navbar {
  display: grid;
  grid-template-rows: 1fr 2.75fr;
  position: fixed;
  top: 0;
  left: -14rem;
  width: 14rem;
  height: 100%;
  padding-left: 1rem;
  z-index: 3;
  background: $white;
  border-right: 1px solid $light-gray;
  transition: all 0.75s ease-in-out;
}

.navbar.open {
  left: 0;
}

.dark-filter {
  position: absolute;
  top: 0;
  left: calc(14rem);
  width: calc(100vw);
  height: 100%;
  background: rgba(0, 0, 0, 0.5);
  z-index: 0;
}

.hide {
  display: none;
}

.nav-logo {
  margin: 0 1rem;
  width: 70%;
}

.nav-links {
  display: flex;
  flex-direction: column;
  width: 100%;
}

.nav-item {
  padding: 0.75rem 1rem;
  width: 100%;
}

.nav-item:hover {
  background-color: $green;
  color: $white;
}
