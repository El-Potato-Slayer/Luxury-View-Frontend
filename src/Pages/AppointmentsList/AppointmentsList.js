import axios from 'axios';
import { useEffect, useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import Notification from '../../Components/Notification/Notification';
import { filterAppointment } from '../../Redux/actions/appointmentActions';
import SkeletonListAgent from '../../Skeletons/SkeletonListAgent/SkeletonListAgent';
import { displayAppointment } from './helper';

function AppointmentsList() {
  const {
    appointments,
    loading,
  } = useSelector((state) => state.appointmentsReducer);
  const dispatch = useDispatch();
  const [success, setSuccess] = useState();
  const [deleteError, setDeleteError] = useState();
  const deleteAppointment = (id) => {
    axios.delete(`appointments/${id}`, {
      headers: {
        Authorization: `token ${localStorage.getItem('token')}`,
      },
    })
      .then(() => {
        dispatch(filterAppointment(id));
        setSuccess('Appointment successfully canceled');
      })
      .catch(() => {
        setDeleteError('An error occurred with the cancelation of the appointment. Please try again later or contact your agent');
      });
  };

  useEffect(() => {
    console.log(appointments);
  }, [appointments]);

  return (
    <div className="page">
      <h1 data-testid="appointments" className="page-title">Appointments</h1>
      {success && <Notification type="success" message={success} />}
      {deleteError && <Notification type="error" message={deleteError} />}

      <div className="listings">
        {loading && (
          <>
            <SkeletonListAgent />
            <SkeletonListAgent />
            <SkeletonListAgent />
          </>
        )}
        {appointments && appointments.map((appointment) => (
          <div key={appointment.id} className="appointment-card">
            {displayAppointment(appointment, deleteAppointment)}
          </div>
        ))}
      </div>
    </div>
  );
}

export default AppointmentsList;
