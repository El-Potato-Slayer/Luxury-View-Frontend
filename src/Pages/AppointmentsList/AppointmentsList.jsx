import axios from 'axios';
import { useEffect, useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import Notification from '../../Components/Notification/Notification';
import {
  fetchAppointmentsFailure, fetchAppointmentsRequest, fetchAppointmentsSuccess, filterAppointment,
} from '../../Redux/actions/appointmentActions';
import SkeletonListAgent from '../../Skeletons/SkeletonListAgent/SkeletonListAgent';
import { displayAppointment } from './helper';
import ConfirmationForm from '../../Components/ConfirmationForm/ConfirmationForm';
import Backdrop from '../../Components/Backdrop/Backdrop';

function AppointmentsList() {
  const {
    appointments,
    loading,
  } = useSelector((state) => state.appointmentsReducer);
  const dispatch = useDispatch();
  const [success, setSuccess] = useState();
  const [deleteError, setDeleteError] = useState();
  const [selectedId, setSelectedId] = useState();
  const [isConfirmOpen, setisConfirmOpen] = useState(false);

  const toggleConfirmation = () => {
    setisConfirmOpen(!isConfirmOpen);
  };

  const deleteAppointment = (id) => {
    axios.delete(`appointments/${id}`, {
      headers: {
        Authorization: `token ${localStorage.getItem('token')}`,
      },
    })
      .then(() => {
        dispatch(filterAppointment(id));
        setSuccess('Appointment successfully canceled');
      })
      .catch(() => {
        setDeleteError('An error occurred with the cancelation of the appointment. Please try again later or contact your agent');
      });
    toggleConfirmation();
  };

  useEffect(() => {
    if (!appointments.length) {
      dispatch(fetchAppointmentsRequest());
      axios.get('appointments', {
        headers: {
          Authorization: `token ${localStorage.getItem('token')}`,
        },
      })
        .then((res) => {
          dispatch(fetchAppointmentsSuccess(res.data));
        })
        .catch(() => {
          dispatch(fetchAppointmentsFailure('Appointments could not be fetched'));
        });
    }
  }, []);

  useEffect(() => {
    const body = document.querySelector('body');
    if (isConfirmOpen) {
      body.style.overflow = 'hidden';
    } else {
      body.style.overflow = 'auto';
    }
  }, [isConfirmOpen]);

  return (
    <>
      {success && <Notification type="success" message={success} />}
      {deleteError && <Notification type="error" message={deleteError} />}
      <div className="page">
        <h1 data-testid="appointments" className="page-title">Appointments</h1>

        {!appointments.length && !loading && (
        <p className="appointments__empty-notification">You dont have any appointments scheduled</p>
        )}
        <div className="listings">
          {loading && (
          <>
            <SkeletonListAgent />
            <SkeletonListAgent />
            <SkeletonListAgent />
          </>
          )}
          {appointments && appointments.map((appointment) => (
            <div key={appointment.id} className="appointment-card">
              {displayAppointment(appointment, setSelectedId, toggleConfirmation)}
            </div>
          ))}
        </div>
        <Backdrop isOpen={isConfirmOpen} formToggler={toggleConfirmation} />
        <ConfirmationForm
          message="Are you sure you want to cancel this appointment?"
          isOpen={isConfirmOpen}
          confirmHandler={() => deleteAppointment(selectedId)}
          rejectHandler={toggleConfirmation}
        />
      </div>
    </>
  );
}

export default AppointmentsList;
