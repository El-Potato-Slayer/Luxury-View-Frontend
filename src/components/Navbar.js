import { useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { Link } from 'react-router-dom';
import { clearUser } from '../store/actions/userActions';
// import { setAuth, setUser } from '../store/actions';

function Navbar() {
  const [burgerState, setBurgerState] = useState(false);
  const { isLoggedIn } = useSelector((state) => state.userReducer);
  const dispatch = useDispatch();
  const logout = () => {
    localStorage.clear();
    dispatch(clearUser());
  };

  const handleBurgerState = () => {
    setBurgerState(!burgerState);
  };

  function authNavItem(isLoggedIn) {
    if (isLoggedIn) {
      return (
        <Link onClick={() => { logout(); handleBurgerState(); }} data-testid="navbar-logout" className="nav-item" to="/">Logout</Link>
      );
    }
    return (
      <>
        <Link onClick={handleBurgerState} data-testid="navbar-login" className="nav-item" to="/login">Login</Link>
        <Link onClick={handleBurgerState} data-testid="navbar-register" className="nav-item" to="/register">Register</Link>
      </>
    );
  }

  return (
    <>
      <div data-testid="navbar" className={`navbar${burgerState ? ' open' : ''}`}>
        <div aria-hidden="true" onClick={handleBurgerState} onKeyDown={handleBurgerState} className={`dark-filter${!burgerState ? ' hide' : ''}`} />
        <div className={`burger${burgerState ? ' open' : ''}`}>
          <button type="button" id="burger-wrapper" className="burger-wrapper" onClick={handleBurgerState}>
            <div id="burger-menu" className={`burger-menu${burgerState ? ' open' : ''}`} />
          </button>
        </div>
        <img src="/assets/logo.png" className="nav-logo" alt="" />
        <div className="nav-links">
          <Link onClick={handleBurgerState} data-testid="navbar-homepage" className="nav-item" to="/">Home</Link>
          <Link onClick={handleBurgerState} data-testid="navbar-mansions" className="nav-item" to="/mansions">Mansions</Link>
          <Link onClick={handleBurgerState} data-testid="navbar-appointments" className="nav-item" to="/appointments">Appointments</Link>
          <Link onClick={handleBurgerState} data-testid="navbar-agents" className="nav-item" to="/agents">Agents</Link>
          {authNavItem(isLoggedIn)}
        </div>
        <div className="nav-social" />
      </div>
    </>
  );
}

export default Navbar;
