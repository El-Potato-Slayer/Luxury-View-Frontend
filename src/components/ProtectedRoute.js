import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Redirect, Route } from 'react-router-dom';

function ProtectedRoute({ isAuth, component: Component, exact }) {
  return (
    <Route
      exact={exact}
      render={(props) => {
        if (isAuth) {
          return <Component />;
        }
        return <Redirect to={{ pathname: '/login', state: { from: props.location } }} />;
      }}
    />
  );
}

ProtectedRoute.propTypes = {
  isAuth: PropTypes.bool,
  component: PropTypes.instanceOf(Object).isRequired,
  exact: PropTypes.bool,
  location: PropTypes.instanceOf(Object),
};

ProtectedRoute.defaultProps = {
  isAuth: false,
  exact: false,
  location: {},
};

// ProtectedRoute.de
// export default ProtectedRoute;
const mapStateToProps = (state) => ({
  isAuth: state.authReducer.isAuth,
});

export default connect(mapStateToProps)(ProtectedRoute);
