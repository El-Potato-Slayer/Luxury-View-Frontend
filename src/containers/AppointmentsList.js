import PropTypes from 'prop-types';
import axios from 'axios';
import { connect, useDispatch } from 'react-redux';
import { useEffect, useState } from 'react';
import { Link } from 'react-router-dom';
import { setAppointments } from '../store/actions';

function AppointmentsList({ appointments }) {
  const dispatch = useDispatch();
  const [currentAppointments, setCurrentAppointments] = useState(appointments);
  const [success, setSuccess] = useState();
  const [deleteError, setDeleteError] = useState();
  const [fetchError, setFetchError] = useState();
  function fetchAppointments() {
    axios.get('appointments', {
      headers: {
        Authorization: `token ${localStorage.getItem('token')}`,
      },
    })
      .then((res) => {
        dispatch(setAppointments(res.data));
        setCurrentAppointments(res.data);
      })
      .catch(() => {
        setFetchError('Appointments could not be fetched. Please try again later');
      });
  }

  const deleteAppointment = (id) => {
    axios.delete(`appointments/${id}`, {
      headers: {
        Authorization: `token ${localStorage.getItem('token')}`,
      },
    })
      .then(() => {
        setCurrentAppointments(
          (currentAppointments) => currentAppointments.filter(
            (appointment) => id !== appointment.id,
          ),
        );
        setSuccess('Appointment successfully deleted');
      })
      .catch(() => {
        setDeleteError('An error occurred with the removal of the appointment. Please try again later or contact your agent');
      });
  };

  function showSuccess(success) {
    if (success) {
      return (
        <p>{success}</p>
      );
    }
    return null;
  }

  function showDeleteError(error) {
    if (error) {
      return (
        <p>{error}</p>
      );
    }
    return null;
  }

  function userButtons(error, id) {
    if (!error) {
      return (
        <button type="button" onClick={() => deleteAppointment(id)}>Cancel</button>
      );
    }

    return null;
  }

  function formatDate(date) {
    const d = new Date(date);
    let arr = d.toString().split(' ').splice(0, 5);
    const time = arr[4].split(':').splice(0, 2).join(':');
    arr = arr.splice(0, 4);
    return `Date: ${arr.join(' ').concat(` ${time}`)}`;

    // return date.toLocaleFormat('MMMM d, yyyy h:mm aa');
  }

  function displayAppointment(appointment) {
    return (
      <div>
        <div className="listing-image" style={{ background: `url(${appointment.property.picture}) center / cover` }} />
        <div className="info">
          <p className="listing-important">{formatDate(appointment.date)}</p>
          <p>
            Property:
            {' '}
            <Link
              to={`mansions/${appointment.property.id}`}
              className="text-info"
            >
              {appointment.property.name}
            </Link>
          </p>
        </div>
        <div className="user-buttons">
          <Link
            to={`appointments/${appointment.id}`}
          >
            Appointment Details
          </Link>
          {userButtons(fetchError, appointment.id)}
        </div>
      </div>
    );
  }

  useEffect(() => {
    fetchAppointments();
  }, []);

  return (
    <div className="page">
      <h1 data-testid="appointments" className="page-title">Appointments</h1>
      {showSuccess(success)}
      {showDeleteError(deleteError)}
      <div className="listings">
        {currentAppointments.map((appointment) => (
          <div key={appointment.id} className="appointment-card">
            {displayAppointment(appointment)}
          </div>
        ))}
      </div>
    </div>
  );
}

const mapStateToProps = (state) => ({
  appointments: state.appointmentsReducer.appointments,
});

AppointmentsList.propTypes = {
  appointments: PropTypes.instanceOf(Array).isRequired,
};

export default connect(mapStateToProps)(AppointmentsList);
