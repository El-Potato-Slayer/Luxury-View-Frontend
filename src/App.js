import './App.css';
import {
  BrowserRouter, Redirect, Route, Switch,
} from 'react-router-dom';
import { connect, useDispatch } from 'react-redux';
import { useEffect } from 'react';
import axios from 'axios';
import PropTypes from 'prop-types';
import Navbar from './components/Navbar';
import Home from './components/Home';
import { setAgents, setAuth, setProperties } from './store/actions';
import MansionsList from './containers/MansionsList';
import AgentsList from './containers/AgentsList';
import ProtectedRoute from './components/ProtectedRoute';
import AppointmentsList from './containers/AppointmentsList';
import Login from './components/Login';
import Mansion from './components/Mansion';

function App({ isAuth }) {
  const dispatch = useDispatch();
  let token;
  useEffect(() => {
    axios.get('http://localhost:3000/api/v1/properties')
      .then((resp) => {
        dispatch(setProperties(resp.data));
      });
    axios.get('http://localhost:3000/api/v1/agents')
      .then((resp) => {
        dispatch(setAgents(resp.data));
      });
    token = localStorage.getItem('token');
    if (token) {
      const tokenExpiration = JSON.parse(atob(token.split('.')[1])).exp;
      const dateNow = new Date();
      if (tokenExpiration > dateNow.getTime() / 1000) {
        dispatch(setAuth(true));
      }
    }
  }, []);
  return (
    <div data-testid="app" className="App">
      <BrowserRouter>
        <Navbar />
        <Switch>
          <Route exact path="/">
            <Home />
          </Route>
          <Route exact path="/mansions">
            <MansionsList />
          </Route>
          <Route exact path="/agents">
            <AgentsList />
          </Route>
          <Route path="/login">
            {isAuth ? <Redirect to="/appointments" /> : <Login />}
          </Route>
          <Route path="/mansions/:id">
            <Mansion />
          </Route>
          <ProtectedRoute exact path="/appointments" component={AppointmentsList} />
        </Switch>
      </BrowserRouter>
    </div>
  );
}

const mapStateToProps = (state) => ({
  isAuth: state.authReducer.isAuth,
});

App.propTypes = {
  isAuth: PropTypes.bool.isRequired,
};

export default connect(mapStateToProps)(App);
