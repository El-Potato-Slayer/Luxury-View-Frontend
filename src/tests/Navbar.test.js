import {
  cleanup, fireEvent, render, screen,
} from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter, Switch } from 'react-router-dom';
import renderer from 'react-test-renderer';
import App from '../App';
import Navbar from '../components/Navbar';
import ProtectedRoute from '../components/ProtectedRoute';
import AppointmentsList from '../containers/AppointmentsList';
import store from '../mocks/store/configureStore';
import authorizedStore from '../mocks/authstore/configureStore';

afterEach(cleanup);

test('Checks if "Home" link contains the text "Home"', () => {
  render(
    <Provider store={store}>
      <BrowserRouter>
        <Navbar />
      </BrowserRouter>
    </Provider>,
  );

  expect(screen.getByTestId('navbar-mansions')).toHaveTextContent('Mansions');
});

test('Checks if "Mansions" link contains the text "Mansions"', () => {
  render(
    <Provider store={store}>
      <BrowserRouter>
        <Navbar />
      </BrowserRouter>
    </Provider>,
  );

  expect(screen.getByTestId('navbar-mansions')).toHaveTextContent('Mansions');
});

test('Checks if "Appointments" link contains the text "Appointments"', () => {
  render(
    <Provider store={store}>
      <BrowserRouter>
        <Navbar />
      </BrowserRouter>
    </Provider>,
  );

  expect(screen.getByTestId('navbar-appointments')).toHaveTextContent('Appointments');
});

test('Checks if "Agents" link contains the text "Agents"', () => {
  render(
    <Provider store={store}>
      <BrowserRouter>
        <Navbar />
      </BrowserRouter>
    </Provider>,
  );

  expect(screen.getByTestId('navbar-agents')).toHaveTextContent('Agents');
});

test('AppointmentsList does not render if unauthorized', () => {
  render(
    <Provider store={store}>
      <App>
        <BrowserRouter>
          <Switch>
            <ProtectedRoute path="/appointments" component={AppointmentsList} isAuth={false} location="/" />
          </Switch>
        </BrowserRouter>
      </App>
    </Provider>,
  );
  fireEvent.click(screen.getByTestId('navbar-appointments'));
  expect(screen.queryByText('Date')).not.toBeInTheDocument();
});

test('Login component renders if unauthorized', () => {
  render(
    <Provider store={store}>
      <App>
        <BrowserRouter>
          <Switch>
            <ProtectedRoute path="/appointments" component={AppointmentsList} isAuth={false} location="/" />
          </Switch>
        </BrowserRouter>
      </App>
    </Provider>,
  );
  fireEvent.click(screen.getByTestId('navbar-appointments'));
  expect(screen.getByTestId('login')).toHaveTextContent('Login');
});

test('AppointmentsList renders if authorized', () => {
  render(
    <Provider store={authorizedStore}>
      <App>
        <BrowserRouter>
          <ProtectedRoute path="/appointments" component={AppointmentsList} isAuth />
        </BrowserRouter>
      </App>
    </Provider>,
  );
  expect(screen.getByTestId('appointments')).toHaveTextContent('Appointments');
});

it('matches snapshot', () => {
  const tree = renderer.create(
    <Provider store={store}>
      <App>
        <BrowserRouter>
          <Navbar />
        </BrowserRouter>
      </App>
    </Provider>,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
