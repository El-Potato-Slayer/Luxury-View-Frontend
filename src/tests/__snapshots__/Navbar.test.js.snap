// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="App"
  data-testid="app"
>
  <div
    className="navbar"
    data-testid="navbar"
  >
    <div
      aria-hidden="true"
      className="dark-filter hide"
      onClick={[Function]}
      onKeyDown={[Function]}
    />
    <div
      className="burger"
    >
      <button
        className="burger-wrapper"
        id="burger-wrapper"
        onClick={[Function]}
        type="button"
      >
        <div
          className="burger-menu"
          id="burger-menu"
        />
      </button>
    </div>
    <img
      alt=""
      className="nav-logo"
      src="/assets/logo.png"
    />
    <div
      className="nav-links"
    >
      <a
        className="nav-item"
        data-testid="navbar-homepage"
        href="/"
        onClick={[Function]}
      >
        Home
      </a>
      <a
        className="nav-item"
        data-testid="navbar-mansions"
        href="/mansions"
        onClick={[Function]}
      >
        Mansions
      </a>
      <a
        className="nav-item"
        data-testid="navbar-appointments"
        href="/appointments"
        onClick={[Function]}
      >
        Appointments
      </a>
      <a
        className="nav-item"
        data-testid="navbar-agents"
        href="/agents"
        onClick={[Function]}
      >
        Agents
      </a>
      <a
        className="nav-item"
        data-testid="navbar-login"
        href="/login"
        onClick={[Function]}
      >
        Login
      </a>
      <a
        className="nav-item"
        data-testid="navbar-register"
        href="/register"
        onClick={[Function]}
      >
        Register
      </a>
    </div>
    <div
      className="nav-social"
    />
  </div>
  <div
    className="form-wrapper"
    data-testid="login"
  >
    <form
      className="form"
      onSubmit={[Function]}
    >
      
      <h2
        className="page-title"
      >
        Login
      </h2>
      <fieldset>
        <input
          className="input-field"
          name="username"
          onChange={[Function]}
          placeholder="Username"
          type="text"
        />
      </fieldset>
      <fieldset>
        <input
          className="input-field"
          name="password"
          onChange={[Function]}
          placeholder="Password"
          type="password"
        />
      </fieldset>
      <button
        className="info-button"
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</div>
`;
